### PR TITLE
Handle extern PyVarObjects

### PR DESCRIPTION
**Pull request by da-woods**
Fixes https://github.com/cython/cython/issues/4827 at least
partially.

Some of patch copied from https://src.fedoraproject.org/rpms/Cython/pull-request/35#request_diff

Allows the size of a type to be between basicsize and basicsize+itemsize
since anything is this range is a reasonable size for a class to be.

Adds a test when an extern extension type is inherited from to
make sure that it isn't a PyVarObject.

### CODE REVIEW

**Comment by da-woods** [comment]
https://foss.heptapod.net/pypy/pypy/-/issues/3772 PyPy bug reported here

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -5457,6 +5458,22 @@ def generate_type_ready_code(entry, code, bases_tuple_cname=None, check_heap_typ
             ))
             code.putln("#endif")  # if CYTHON_USE_TYPE_SPECS
 
+            base_type = type.base_type
+            while base_type:
+                if type.base_type.is_external and not base_type.objstruct_cname == "PyTypeObject":
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:5463]: Did you really mean `type.base_type` here or just `base_type`?

File: tests/run/extern_varobject_extensions.srctree [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+check_tuple((1, 2))
+check_list([1, 2])
+check_code(eval("lambda: None").__code__)
+
+##### failed_inherit.pyx #############
+
+from classes cimport tpl
+
+cdef class SuperTuple(tpl):
+    cdef int a  # importing this gives an error message
```
Comment (scoder) [inline tests/run/extern_varobject_extensions.srctree:61]: We should also test only adding a cdef method, without additional struct attributes.

File: Cython/Utility/ImportExport.c [hunk 3]
```diff
@@ -522,19 +525,30 @@ static PyTypeObject *__Pyx_ImportType(PyObject *module, const char *module_name,
...
             "Expected %zd from C header, got %zd from PyObject",
-            module_name, class_name, size, basicsize);
+            module_name, class_name, size, basicsize+itemsize);
         goto bad;
     }
-    if (check_size == __Pyx_ImportType_CheckSize_Error && (size_t)basicsize != size) {
+    // varobjects almost have structs  between basicsize and basicsize + itemsize
+    // but the struct isn't always one of the two limiting values
+    if (check_size == __Pyx_ImportType_CheckSize_Error &&
+            ((size_t)basicsize > size || (size_t)(basicsize + itemsize) < size)) {
```
Comment (scoder) [inline Cython/Utility/ImportExport.c:547]: I don't think this assumption is correct. `itemsize` gives the size of one item. If there are more items, then this miscalculates the size.

File: tests/run/extern_varobject_extensions.srctree [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+from Cython.Build.Dependencies import cythonize
+
+from distutils.core import setup, Extension
+
+setup(
+  ext_modules=cythonize([
+    Extension("classes", ["classes.pyx"]),
+    Extension("failed_inherit", ["failed_inherit.pyx"]),
+    Extension("successful_inherit", ["successful_inherit.pyx"]),
+  ]),
```
Comment (scoder) [inline tests/run/extern_varobject_extensions.srctree:19]:
```suggestion
  ext_modules=cythonize("*.pyx"),
```

File: Cython/Utility/ImportExport.c (hunk 3 above)
Comment (da-woods) [reply Cython/Utility/ImportExport.c:547]: Unless I'm missing the point of the comment, the varobject struct is almost inevitably:

```
struct {
    ...
    ItemType items[0];
};
```
or

```
struct {
    ...
    ItemType items[1];
};
```
or
```
struct {
    ...
    ItemType items[];
};
```
So the calculation of "is the struct size right" is based on there either being 0 or 1 items. Obviously extra space may be allocated at runtime for more items, but doesn't need considering here

Comment (da-woods) [reply Cython/Utility/ImportExport.c:547]: It definitely can be wrong though - I think there should be an option to completely ignore the size check or to specify an expected size. But I don't think it needs to be part of this change

**Comment by scoder** [comment]
Ok, I think it's a net improvement that reduces false positives of the size check. Let's get this into the next alpha.

File: tests/run/extern_varobject_extensions.srctree [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+PYTHON setup.py build_ext --inplace
+PYTHON -c "import classes"
+PYTHON -c "import test_inherit"
+
+######## setup.py ########
+
+from Cython.Build.Dependencies import cythonize
+
+from distutils.core import setup, Extension
```
Comment (scoder) [inline tests/run/extern_varobject_extensions.srctree:11]:
```suggestion
from distutils.core import setup
```

**Comment by scoder** [comment]
Hmm. I should have checked the test results more closely. Now I'm getting this in Py3.11b4:
```
Errors from shard 2:
======================================================================
FAIL: runTest (__main__.EndToEndTest.runTest)
[2] End-to-end extern_varobject_extensions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/cython/cython/runtests.py", line 1969, in runTest
    self.assertEqual(0, res, "non-zero exit status, last output was:\n%r\n-- stdout:%s\n-- stderr:%s\n" % (
AssertionError: 0 != 1 : non-zero exit status, last output was:
'/opt/hostedtoolcache/Python/3.11.0-beta.4/x64/bin/python -c import classes'
-- stdout:
-- stderr:Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "classes.pyx", line 2, in init classes
    def check_tuple(tpl x):
ValueError: types.CodeType size changed, may indicate binary incompatibility. Expected 192 from C header, got 186 from PyObject
```

**Comment by scoder** [comment]
It's failing on this condition now:
```c
    if ((size_t)(basicsize + itemsize) < size) {
        PyErr_Format(PyExc_ValueError,
            "%.200s.%.200s size changed, may indicate binary incompatibility. "
            "Expected %zd from C header, got %zd from PyObject",
            module_name, class_name, size, basicsize);
            module_name, class_name, size, basicsize+itemsize);
        goto bad;
    }
```

**Comment by da-woods** [comment]
It looks like Py3.11 `CodeType` doesn't quite fit into this size scheme (which seems like a problem because it's what the original issue was about). I guess the options are:
1. Exclude `CodeType` from the tests - we haven't yet got a good solution to handle it, but this PR is still better than what was there before,
2. Revert this change
3. Add a bit of extra wiggle-room in the sizing - let `sizeof(struct)` be up to `itemsize` larger than `basicsize` to handle the case where `basicsize` is defined by `offsetof(final_struct_attribute)`. That makes it a weaker guard but would probably cover `CodeType`

**Comment by scoder** [comment]
Could that be an alignment issue? We have this at the end of the struct:
```
    char *_co_linearray;          /* array of line offsets */                  \
    int _co_firsttraceable;       /* index of first traceable instruction */   \
    /* Scratch space for extra data relating to the code object.               \
       Type is a void* to keep the format private in codeobject.c to force     \
       people to go through the proper APIs. */                                \
    void *co_extra;                                                            \
    char co_code_adaptive[(SIZE)];                                             \
```
So there are two pointers around an `int`.

**Comment by da-woods** [comment]
I think there's a fairly substantial amount of padding after co_code_adaptive. It looks like `_Alignof(PyCodeObject)` is 8 (at least on 64bit linux), so I guess the size is rounded to the nearest 8

**Comment by da-woods** [comment]
Possibly suggests that we should take itemsize as either 0 (for `tp_itemsize==0`), or the largest of `tp_itemsize` or `alignof(struct)`.

Although `alignof` is a C++11/C11 feature, so we'd probably have to fall back to `sizeof(void*)` or earlier standards.

**Pull request merged**